[KeyInstr] Complex assignment atoms

OCHyams opened the pull request:
This patch is part of a stack that teaches Clang to generate Key Instructions
metadata for C and C++.

The Key Instructions project is introduced, including a "quick summary" section
at the top which adds context for this PR, here:
https://discourse.llvm.org/t/rfc-improving-is-stmt-placement-for-better-interactive-debugging/82668

The feature is only functional in LLVM if LLVM is built with CMake flag
LLVM_EXPERIMENTAL_KEY_INSTRUCTIONs. Eventually that flag will be removed.

The Clang-side work is demoed here:
https://github.com/llvm/llvm-project/pull/130943

OCHyams committed:
[KeyInstr] Complex assignment atoms

This patch is part of a stack that teaches Clang to generate Key Instructions
metadata for C and C++.

The Key Instructions project is introduced, including a "quick summary" section
at the top which adds context for this PR, here:
https://discourse.llvm.org/t/rfc-improving-is-stmt-placement-for-better-interactive-debugging/82668

The feature is only functional in LLVM if LLVM is built with CMake flag
LLVM_EXPERIMENTAL_KEY_INSTRUCTIONs. Eventually that flag will be removed.

The Clang-side work is demoed here:
https://github.com/llvm/llvm-project/pull/130943

OCHyams commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134638?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#134640** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134640?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#134639** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134639?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#134638** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134638?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134638?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#134637** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134637?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#134636** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134636?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#134635** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134635?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#134634** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134634?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#134633** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134633?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#134632** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134632?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#134627** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/134627?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133495** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133495?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133494** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133494?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133493** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133493?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133492** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133492?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133491** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133491?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133490** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133490?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133489** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133489?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133488** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133488?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133487** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133487?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133486** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133486?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133485** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133485?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133484** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133484?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133483** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133483?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133482** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133482?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133481** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133481?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133480** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133480?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133479** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133479?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133478** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133478?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#133477** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/133477?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

OCHyams committed:
cc1

OCHyams committed:
whitespace

jmorse commented on clang/lib/CodeGen/CGExprComplex.cpp:
```diff
@@ -1209,6 +1213,7 @@ LValue ComplexExprEmitter::
 EmitCompoundAssignLValue(const CompoundAssignOperator *E,
           ComplexPairTy (ComplexExprEmitter::*Func)(const BinOpInfo&),
                          RValue &Val) {
+  ApplyAtomGroup Grp(CGF.getDebugInfo());
```
This doesn't appear to be symmetrical with the scalar patch, in that `EmitCompoundAssign` gets an atom group for the whole assignment -- should this be the same?

OCHyams replied:
>  in that `EmitCompoundAssign` gets an atom group for the whole assignment

I don't think that's true?

jmorse commented on clang/lib/CodeGen/CGExprComplex.cpp:
```diff
@@ -1356,6 +1361,7 @@ LValue ComplexExprEmitter::EmitBinAssignLValue(const BinaryOperator *E,
 ComplexPairTy ComplexExprEmitter::VisitBinAssign(const BinaryOperator *E) {
+  ApplyAtomGroup Grp(CGF.getDebugInfo());
   ComplexPairTy Val;
   LValue LV = EmitBinAssignLValue(E, Val);
```
Similarly, if we followed the pattern of the above hunk, we'd put ApplyAtomGroup into `EmitBinAssignLValue` instead. This one covers the load-of-lvalue at the end, wheras for compound assignments the equivalent wouldn't be covered.

OCHyams replied:
I'm a bit confused, the previous patch puts the ApplyAtomGroup in `ScalarExprEmitter::VisitBinAssign`, so this looks symmetrical to me - am I missing something?

jmorse reviewed (commented):
Looks good overall but I have some minor comments about symmetry with other patches

OCHyams commented:
Sorry to all that were added... grappling with messed-up graphite internal state, GitHub enjoys adding reviewers for me when I change base.

OCHyams committed:
expand test

OCHyams committed:
further expand test

OCHyams committed:
shuffle code, works for cpp

OCHyams committed:
fix c scalar bin-op-assign complex

OCHyams committed:
fix c scalar binop assign complex unecessary atom nesting

OCHyams commented:
Sorry I misunderstood what you were saying. Fixed, this should now have symmetry with prior patches. Expanded the test coverage. There's 4 binop assign calls that are instrumented individually which reflects the previous patch's macro usage. The expanded test coverage identified a redundant atom group nesting, which is now also fixed (and could only be fixed with the symmetrical approach, hurray). Thanks for the keen eyed review!

jmorse reviewed (approved):
LGTM

OCHyams committed:
add triple to test

OCHyams committed:
remove now-incorrect coment
